### PR TITLE
[FLOC-2735] Persist leases

### PR DESCRIPTION
**Pull request by itamarst**
1. Leases are now a persistent piece of information in the configuration.
2. Leases can expire, in which case they are removed from the configuration.
3. The configuration format has been updated to address this change.

### CODE REVIEW

File: flocker/control/test/test_persistence.py [hunk 1]
```diff
@@ -57,56 +64,100 @@
...
-        d.addCallback(check_expired)
+        def saved(_):
+            logger.reset()
+            self.clock.advance(1000)
+            assertHasMessage(self, logger, _LOG_EXPIRE, {
+                u"dataset_id": dataset_id, u"node_id": node_id})
+        d.addCallback(saved)
         return d
 
-
```
Comment (jml) [inline flocker/control/test/test_persistence.py]: This change is causing the lint test to fail

http://build.clusterhq.com/builders/flocker-lint/builds/9364

**Comment by jml** [comment]
There are a bunch of failures in `flocker.control.test.test_protocol` that look legitimate.

File: flocker/control/test/test_persistence.py [hunk 2]
```diff
@@ -363,6 +415,13 @@ def test_sequential_migrations(self):
 
 DATASETS = st.builds(Dataset, dataset_id=UUIDS, maximum_size=st.integers())
 
+LEASES = st.builds(
+    Lease, dataset_id=UUIDS, node_id=UUIDS,
+    expiration=st.one_of(
+        st.none(),
+        st.basic(generate=lambda r, _: datetime.fromtimestamp(
+            r.randrange(0, 10000000), tz=UTC))))
```
Comment (jml) [inline flocker/control/test/test_persistence.py]: I suspect `integers(max_value=10000000).map(lambda t: datetime.fromtimestamp(t, tz=UTC))` would perform better. It's cleaner too.

**Comment by jml** [comment]
Hey Itamar,

It's pretty cool seeing this after all our design work in Cambridge.

I've done a review for clarity rather than correctness[1], and only have minor comments. The tests, however, are failing and maybe for good reasons? Anyway, if major changes are required, please resubmit for review.

jml

**Comment by dwgebler** [comment]
I have added a commit to fix the failing `flocker.control.test.test_protocol` and the lint error.

**Comment by jml** [comment]
:+1: LGTM. Merge when the tests pass.

**Pull request merged**